File: array/69_Sqrt/69.cpp
class Solution {
public:
    int mySqrt(int x) {
        if (x < 2) return x;
        long left = 1, right = x / 2, middle;
        while (left <= right) {
            middle = left + ((right - left) >> 1);
            if (middle * middle < x) {
               left = middle + 1;
            } else if (middle * middle > x) {
                right = middle - 1;
            } else {
                return middle;
            }
        }
        return right;
    }
};

File: array/69_Sqrt/69.py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x < 2:
            return x
        left, right = 1, x // 2
        while left <= right:
            middle = left + ((right - left) >> 1)
            if middle * middle < x:
                left = middle + 1
            elif middle * middle > x:
                right = middle - 1
            else:
                return middle
        return right
